plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2229,9 +2228,27 @@ private void finishWrite(ConnectorSession session, IcebergTableHandle table, Col
 
         try {
             if (!fullyDeletedFiles.isEmpty()) {
-                DeleteFiles deleteFiles = transaction.newDelete();
-                fullyDeletedFiles.keySet().forEach(deleteFiles::deleteFile);
-                commit(deleteFiles, session);
+                OverwriteFiles overwriteFiles = transaction.newOverwrite();
+                TableScan tableScan = catalog.loadTable(session, table.getSchemaTableName()).newScan();
+                table.getSnapshotId().ifPresent(tableScan::useSnapshot);
+                try (CloseableIterable<FileScanTask> taskPlan = tableScan.planFiles(); CloseableIterator<FileScanTask> taskIterator = taskPlan.iterator()) {
```

Comment:
@danielcweeks do you know if there's a better way to go from a file path to a DataFile? I'd rather not have to go through a whole table scan plan again here

Reply:
You might be able to build it from `DataFiles.Builder`.  There is some required information like path, size, record count, which I think you may have in context.  That might be enough to avoid the full metadata scan.

Reply:
It's definitely a bad idea to scan the table during job commit. I think you should pass the DataFile all the way through to this point. Is that not possible for some reason?

Reply:
I think the original reason for not doing that has to do with the fact that DataFiles are not Jackson serializable. We get around that for some of the Iceberg classes, like Schema and PartitionSpec, with their JSON parsers and serialize them to Strings. Is there one of those for DataFiles?

Reply:
There isn't a parser for DataFile, but it is a well-defined interface so you could probably build one fairly easily. I can review if you want.

Reply:
You can also use Java serialization and push that into a JSON string if you need to and don't want to wait for a parser.